doozerlib/cli/get_nightlies.py

```diff
@@ -107,16 +219,277 @@ def rc_api_url(tag, arch):
+                    # most of the time they'll have the same VR but that's not absolutely
+                    # guaranteed (one build could flake then succeed with later R).
+                    # so just rely on source commit equivalence (already verified)
+                    # and ignore the slim possibility that the RPMs installed differ.
+                    continue
+                logger.debug(f"different because {tag} NVR {self_nvr} != {other_nvr}")
+                return False
+
+        return self.deeper_equivalence_rhcos(other)
+
+    def deeper_equivalence_rhcos(self, other: 'EquivalenceBase') -> bool:
+        """Check that the two have the same RHCOS contents according to build records"""
+        for eq_base in (self, other):
+            if not eq_base.rhcos_inspector:
+                raise Exception(f"No rhcos_inspector for eq_base {eq_base}, should have called populate_deeper_equivalence first")
+            eq_base._rhcos_rpms = {
+                nevra[0]: (nevra[2], nevra[3])
+                for nevra in eq_base.rhcos_inspector.get_os_metadata_rpm_list()
+            }
+
+        logger.debug(f"comparing {self.rhcos_inspector} and {other.rhcos_inspector}")
+        for rpm_name, vr in self._rhcos_rpms.items():
+            if rpm_name in other._rhcos_rpms and vr != other._rhcos_rpms[rpm_name]:
+                logger.debug(f"different '{rpm_name}' version-release {vr} != {other._rhcos_rpms[rpm_name]}")
+                return False
+
+        return True
+
+
+class EqSetDuplicateArchException(Exception):
+    """Indicates a bug where code tried to add a nightly with an arch that's already represented in the eq set"""
+
+
+class EquivalenceSet:
+    """
+    Class to represent a set of nightlies from each arch that are equivalent
+    according to their EquivalenceBase. They can be sorted by creation date
+    (latest timestamp of nightlies in the set)
+    """
+
+    def __init__(self, nightly_for_arch: Dict[str, Dict]):
+        self.nightly_for_arch = nightly_for_arch
+        self.timestamp = max(nightly["releaseInfo"]["config"]["created"] for nightly in nightly_for_arch.values())
+
+    def generate_equivalents_with(self, arch: str, nightlies: List[Dict]) -> List['EquivalenceSet']:
+        """
+        Test each nightly for equivalence with all existing members, and return a list
+        of new sets extended with those that match.
+        """
+        if arch in self.nightly_for_arch:
+            raise EqSetDuplicateArchException(f"Cannot add arch {arch} when already present in {self}")
+
+        new_sets: List[EquivalenceSet] = []
+        for nightly in nightlies:
+            nightly_for_arch = {arch: nightly}  # initialize new set with candidate nightly
+            for exarch, existing in self.nightly_for_arch.items():
+                logger.debug(f"comparing {nightly['name']} and {existing['name']}")
+                if nightly["equivalence"] != existing["equivalence"]:
```

Comment:
This block looks inverted to me.. any reason why we can't break when we find the first match, like
```python
if nightly["equivalence"] == existing["equivalence"]:
    nightly_for_arch[exarch] = existing
    break
```

Reply:
I'm not sure why we're not returning here when the first matching nightly is found for a set.. are we expecting a case where multiple identical nightlies for an arch could match?